src/inc/til/mutex.h

```diff
@@ -0,0 +1,128 @@
+            {
+                _mutex.lock();
+            }
+
+        public:
+            ~guard() noexcept
+            {
+                _mutex.unlock();
+            }
+
+            guard(const guard&) = delete;
+            guard& operator=(const guard&) = delete;
+
+            T* operator->() noexcept
+            {
+                return &_data;
+            }
+
+        private:
+            // We could reduce this to a single pointer member,
+            // by storing a reference to the outer mutex& class instead
+            // and accessing its private members as a friend class.
+            // The reason we don't do that is because C++ compilers usually
+            // don't "cache" dereferenced pointers between mutations.
+            //
+            // For instance if we had:
+            //   struct foo { int a, b; };
+            //   struct bar { foo& f; };
+            //
+            //   void test(bar& b) {
+            //       b.f.a = 123;
+            //       b.f.b = 456;
+            //   }
+            //
+            // This would generate the following suboptimal assembly despite /O2:
+            //   mov rax, QWORD PTR [rcx]
+            //   mov DWORD PTR [rax], 123
+            //
+            //   mov rax, QWORD PTR [rcx]
+            //   mov DWORD PTR [rax+4], 456
+            T& _data;
+            std::shared_mutex& _mutex;
+        };
+
+        // A shared, read-only reference to a til::shared_mutex's underlying data.
+        // If you drop the shared_guard the mutex is unlocked.
+        class shared_guard
+        {
+            friend class shared_mutex;
+
+            shared_guard(const T& data, std::shared_mutex& mutex) :
+                _data(data),
+                _mutex(mutex)
+            {
+                _mutex.lock_shared();
+            }
+
+        public:
+            ~shared_guard() noexcept
+            {
+                _mutex.unlock_shared();
+            }
+
+            shared_guard(const guard&) = delete;
+            shared_guard& operator=(const shared_guard&) = delete;
+
+            const T* operator->() const noexcept
+            {
+                return &_data;
+            }
+
+        private:
+            const T& _data;
+            std::shared_mutex& _mutex;
+        };
+
+        shared_mutex() = default;
+
+        template<typename... Args>
+        shared_mutex(Args&&... init) :
+            _data{ std::forward<Args>(args)... }
+        {
+        }
+
+        // Acquire an exclusive, read/write reference to T.
+        // For instance:
+        //   .lock()->foo = bar;
+        guard lock() const noexcept
+        {
+            return { _data, _mutex };
+        }
+
+        // Acquire a shared, read-only reference to T.
+        // For instance:
+        //   bar = .lock_shared()->foo;
+        shared_guard lock_shared() const noexcept
+        {
+            return { _data, _mutex };
+        }
+
+    private:
+        mutable T _data{};
```

Comment:
I didn't expect _data to be mutable. Now there's no difference between const and non-const shared_mutex because all data members are mutable and all member functions are const.

Reply:
That's exactly right! This is because this mutex class encapsulates not just the data, but also const-correctness! In Rust this concept is called "interior mutability".

The mutex ensures const-correctness, by enforcing that only a single writer at a time can acquire a non-const reference to the inner (protected) data using a call to `.lock()`. If no routine has currently `.lock()`ed the mutex, multiple const references to the inner data can be created, by calling `.lock_shared()`. That's why `_data` is `mutable`: The mutex itself controls const-correctness.

Reply:
BTW I also tested today whether there's any difference in the generated assembly between using `mutable` and not and there doesn't seem to be any.

But this isn't really surprising anyways: As usual C++ continues to not have any rules forbidding aliasing non-const references, which sucks 🍑. Meaning: Every time you dereference a value in C++ you fetch it's value from memory, because the compiler can't prove you didn't change it as a side-effect somewhere else. In C++ every pointer is something of a 50%-volatine-pointer. So yeah... `mutable` doesn't make it worse, because it's already suboptimal by default lol.

Reply:
Oh, I wasn't thinking about aliasing and code generation.  Rather, I was thinking about having a `til::shared_mutex mutex` data member in some type X.  Even `const` member functions of X can then exclusively lock and mutate the data within `X::mutex`. It seems to me that, if the `const` member functions of `X` need to do that, then having `X::mutex` itself be mutable would make the code easier to understand. I am not familiar with Rust, though.

Reply:
My first comment explains the reasoning why this mutex works even if in const member functions of X.

The gist of it was:
Imagine a programmer accidentally uses `.lock_shared()` in a non-const method and modify the data protected by the mutex! All hell would break loose, because you just created a race condition.
By tying const-ness to the way you locked (exclusive/shared) you can get much better safety instead.
When you use `.lock_shared()` with this mutex you can never get a non-const reference and accidentally modify the data. 

But let's talk about your concern:
* If you use a class instance across threads in, the sharing **must** occur using constant references or a `std::shared_ptr<const T>`. Why? Well if your class has any methods that aren't thread-safe, all such methods must be marked const, otherwise you risk a race condition. To ensure this safety, cross-thread sharing must always use const-ness.... technically. I know the real world works differently, but that's really just an unfortunate of C++ age, isn't it?
* One solution would be to make the `til::shared_mutex` member of the class itself `mutable`, allowing you to call `.lock()`, like you said. But this would run counter to the very idea this class represents: It encapsulates cross-thread shared state and the const-ness of its data, by the way you locked it, and not by whether the caller is const or not. In fact it doesn't matter whether the caller is `const`, because the caller should always be `const` when this class is used (due to the prior point). But yeah, I could add the `mutable` keyword anywhere this class is used instead for sure, but it would certainly be against the concept and ideals this class represents. Especially since many probably agree that the `mutable` keyword is almost as bad as `const_cast` itself. I mean it does the same thing after all. Personally I'd rather have this anti-pattern encapsulated just once, which is why I did it this way...

Reply:
Damn.... Now that I wrote all this I realize a much much shorter answer:
You cannot have any mutable references to a class instance once it's potentially shared across threads.
So why bother making this mutex work with the old C++ concept of "the caller's constness determines the callee constness", when the former must always be const anyways?